octobot_trading/exchanges/traders/trader.py

```diff
@@ -326,46 +326,86 @@ async def cancel_order(self, order: object, ignored_order: object = None,
+                    return True
+            except Exception as e:
+                self.logger.exception(e, True, f"Failed to cancel order {order}")
+                return False
+            if order_status is enums.OrderStatus.CANCELED:
                 order.status = octobot_trading.enums.OrderStatus.CANCELED
+                self.logger.debug(f"Successfully cancelled order {order}")
+            elif order_status is enums.OrderStatus.PENDING_CANCEL:
+                order.status = octobot_trading.enums.OrderStatus.PENDING_CANCEL
+                self.logger.debug(f"Order cancel in progress for {order}")
+        else:
+            order.status = octobot_trading.enums.OrderStatus.CANCELED
 
         await order.on_cancel(force_cancel=order.status is octobot_trading.enums.OrderStatus.CANCELED,
                               is_from_exchange_data=False,
                               ignored_order=ignored_order)
         if wait_for_cancelling and order.state is not None and order.state.is_pending():
-            self.logger.debug(f"Waiting for order cancelling, order: {order}")
-            await order.state.wait_for_terminate(cancelling_timeout)
-            self.logger.debug(f"Completed order cancelling, order: {order}")
+            await self._wait_for_order_cancel(order, cancelling_timeout)
         return True
 
+    async def _handle_order_cancel_error(self, order, err, wait_for_cancelling, cancelling_timeout):
+        # can't cancel order: it likely means that the order is not open on exchange anymore
+        if order.state is None:
+            raise err
+        # trigger forced refresh to get an update of the order
+        if order.state.is_refreshing():
+            await order.state.wait_for_next_state(cancelling_timeout)
+        else:
+            previous_status = order.status
+            await order.state.synchronize(force_synchronization=True)
+            if previous_status != order.status:
+                # status changed: wait for state change
+                await order.state.wait_for_next_state(cancelling_timeout)
+        if order.is_cancelled():
+            self.logger.debug(f"Tried to cancel an already cancelled order.")
+            return True
+        if order.is_cancelling():
+            if wait_for_cancelling:
+                await self._wait_for_order_cancel(order, cancelling_timeout)
+            return True
+        elif order.is_open():
+            raise errors.OpenOrderError(
+                "Order is open, but can't be cancelled. This is unexpected"
+            ) from err
+        elif order.is_filled():
+            raise errors.FilledOrderError("Order is filled, it can't be cancelled") from err
+        elif order.is_closed():
+            raise errors.ClosedOrderError("Order is closed, it can't be cancelled") from err
+        else:
+            # should not happen
+            raise errors.OrderCancelError(
```

Comment:
Can this raise produce unexpected effects? In other words, do we need to catch this error somewhere?

Reply:
this is was happens at the moment (before PR), it will forward the error to the caller. I think we should leave it like this as it's a real problem if it happens

Reply:
I agree. The caller is the order updater right?

Reply:
Here the caller is the thing that called trader.cancel_order(), usually a trading mode or a user command

Reply:
Ok. Today the error is forwarded to the caller but don't you think we should add a catch in trading modes to prevent unexpected behaviors?

Reply:
It's the same as for other unexpected state errors (when the order is filled etc), I think it's more on a trading mode by trading mode basis to decide. I'm not sure we can have a generic thing here. What is your opinion ?

Reply:
I agree for specific calls but for more generic ones like the cancel all at startup I think we should add something